File: golang/architecture-shell-cqrs/unit_of_work.go
package cqrs

import "context"

// UnitOfWork represents transaction boundary abstraction for commands.
//
// Transaction Semantics:
//   - Commands open transactions via BeginTransaction()
//   - Queries do NOT open transactions (BR-003)
//   - Nested commands reuse active transaction (no nested transaction opening)
//   - Commit occurs on successful handler completion (including Result.Failure business errors)
//   - Rollback occurs on error (infrastructure errors only)
//
// Implementation Notes:
//   - Transaction context propagates via context.Context values
//   - BeginTransaction() throws if transaction provider unavailable (fail fast per BR-006)
type UnitOfWork interface {
	// TransactionID returns the unique identifier for the current transaction.
	// Used for correlation in telemetry logs per NFR-002.
	TransactionID() string

	// HasActiveTransaction indicates whether a transaction is currently active.
	// Used by UnitOfWork behavior to detect nested commands and reuse existing transaction.
	HasActiveTransaction() bool

	// BeginTransaction begins a new transaction.
	// Returns error if transaction provider unavailable (connection pool exhaustion, database down).
	// Fail-fast behavior per BR-006.
	BeginTransaction(ctx context.Context) error

	// Commit commits the current transaction.
	// Called on successful handler completion (including Result.Failure business errors per BR-008).
	// Returns error if commit fails (e.g., constraint violation, deadlock).
	Commit(ctx context.Context) error

	// Rollback rolls back the current transaction.
	// Called on error (infrastructure errors, handler panics).
	// Returns error if rollback fails (should be logged but not re-thrown).
	Rollback(ctx context.Context) error
}

File: java-spring/architecture-shell-cqrs/src/main/java/com/architecture/shell/cqrs/UnitOfWork.java
package com.architecture.shell.cqrs;

import java.util.UUID;

/**
 * Transaction boundary abstraction for commands.
 * Manages database transactions with support for nested command detection.
 *
 * Requirements:
 * - BR-001: Commands execute within transaction boundaries
 * - BR-002: TransactionId logged for all command executions (telemetry requirement)
 * - BR-003: Queries do NOT open transactions
 * - BR-005: Nested commands reuse active transaction (no nested transaction opening)
 * - BR-006: Fail fast when transaction provider unavailable (BeginTransaction throws)
 * - BR-007: Result.failure() commits transaction, exceptions rollback
 *
 * Transaction Lifecycle:
 * 1. UnitOfWorkBehavior checks hasActiveTransaction()
 * 2. If no active transaction: beginTransaction()
 * 3. Handler executes within transaction context
 * 4. If Result.failure() or success: commit()
 * 5. If exception thrown: rollback()
 *
 * Nested Command Pattern:
 * <pre>
 * Outer Command:
 *   1. UnitOfWork.beginTransaction() → TransactionId = ABC
 *   2. Handler calls Mediator.send(InnerCommand)
 *
 * Inner Command:
 *   1. UnitOfWork.hasActiveTransaction() → true
 *   2. Skip beginTransaction(), reuse ABC
 *   3. Handler executes in ABC transaction
 *
 * Outer Command (continued):
 *   4. commit() or rollback() based on outer command result
 * </pre>
 *
 * DDD Layer: Application Layer Infrastructure
 *
 * @see UnitOfWorkBehavior
 */
public interface UnitOfWork {

    /**
     * Gets the current transaction ID.
     * Required for telemetry logging per BR-002 and NFR-002.
     *
     * @return The transaction ID, or null if no active transaction
     */
    UUID getTransactionId();

    /**
     * Checks if a transaction is currently active.
     * Used to detect nested commands per BR-005.
     *
     * @return true if transaction is active in current context
     */
    boolean hasActiveTransaction();

    /**
     * Begins a new transaction.
     * MUST be called only when hasActiveTransaction() returns false.
     *
     * @throws IllegalStateException if transaction provider unavailable (BR-006 fail-fast)
     * @throws IllegalStateException if transaction already active (caller violated BR-005 check)
     */
    void beginTransaction();

    /**
     * Commits the current transaction.
     * Called on successful handler completion, including Result.failure() per BR-007/BR-008.
     *
     * @throws IllegalStateException if no active transaction
     * @throws RuntimeException if commit fails (infrastructure error)
     */
    void commit();

    /**
     * Rolls back the current transaction.
     * Called when handler or behavior throws exception per BR-007.
     *
     * @throws IllegalStateException if no active transaction
     * @throws RuntimeException if rollback fails (logged but not rethrown to preserve original exception)
     */
    void rollback();
}
